tests/monadic_expr.x

```diff
@@ -0,0 +1,121 @@
+{
+module Main (main) where
+import {- "containers" -} Data.Set (Set)
+import {- "containers" -} qualified Data.Set as Set
+import {- "text" -} Data.Text (Text)
+import {- "text" -} qualified Data.Text as Text
+import {- "text" -} qualified Data.Text.Read as Text
+import {- "base" -} Control.Arrow hiding (arr)
+import {- "base" -} Control.Monad
+import {- "base" -} Control.Monad.Fail
+import {- "base" -} Numeric.Natural
+import {- "base" -} System.Exit
+}
+
+%wrapper "monadUserState-strict-text"
+%token "Token integer"
+%typeclass "Integral integer, Read integer, Show integer"
+
+-- ugh
+$digit = 0-9
+$unidigit = 1-9
+@number = [0] | $unidigit $digit*
+
+tokens :-
+  $white+ { skip }
+  @number { \(_, _, _, s) len -> case Text.decimal (Text.take len s) of
+                  Left e -> fail e
+                  Right (n, txt)
+                    | Text.null txt -> pure $ TokenInt n
+                    | otherwise -> fail "got incomplete prefix " }
```

Comment:
For GHC 8.4/8.6: `fail` -> `Control.Monad.Fail.fail`  (or hide`fail` from `Control.Monad`).
In general, just using explicit imports makes things more robust across multiple GHC versions...

Reply:
Here is the whole (hi)story: https://gitlab.haskell.org/haskell/prime/-/wikis/libraries/proposals/monad-fail#transitional-strategy

Reply:
It also comes in through `Prelude`; I'm just using it qualified now.